consumer/consumererror/signalerrors.go

```diff
@@ -20,77 +20,90 @@ import (
 	"go.opentelemetry.io/collector/pdata/ptrace"
 )
 
+type retryable[V ptrace.Traces | pmetric.Metrics | plog.Logs] struct {
+	error
+	failed V
+}
+
+// Unwrap returns the wrapped error for functions Is and As in standard package errors.
+func (err retryable[V]) Unwrap() error {
+	return err.error
+}
+
 // Traces is an error that may carry associated Trace data for a subset of received data
 // that failed to be processed or sent.
 type Traces struct {
-	error
-	failed ptrace.Traces
+	retryable[ptrace.Traces]
 }
 
 // NewTraces creates a Traces that can encapsulate received data that failed to be processed or sent.
 func NewTraces(err error, failed ptrace.Traces) error {
 	return Traces{
-		error:  err,
-		failed: failed,
+		retryable: retryable[ptrace.Traces]{
+			error:  err,
+			failed: failed,
+		},
 	}
 }
 
-// GetTraces returns failed traces from the associated error.
+// Deprecated: [v0.72.0] Use `Traces` instead.
 func (err Traces) GetTraces() ptrace.Traces {
-	return err.failed
+	return err.Traces()
 }
 
-// Unwrap returns the wrapped error for functions Is and As in standard package errors.
-func (err Traces) Unwrap() error {
-	return err.error
+// Traces returns failed traces from the associated error.
+func (err Traces) Traces() ptrace.Traces {
+	return err.failed
 }
 
 // Logs is an error that may carry associated Log data for a subset of received data
 // that failed to be processed or sent.
 type Logs struct {
-	error
-	failed plog.Logs
+	retryable[plog.Logs]
 }
 
 // NewLogs creates a Logs that can encapsulate received data that failed to be processed or sent.
 func NewLogs(err error, failed plog.Logs) error {
 	return Logs{
-		error:  err,
-		failed: failed,
+		retryable: retryable[plog.Logs]{
+			error:  err,
+			failed: failed,
+		},
 	}
 }
 
-// GetLogs returns failed logs from the associated error.
+// Deprecated: [v0.72.0] Use `Logs` instead.
 func (err Logs) GetLogs() plog.Logs {
-	return err.failed
+	return err.Logs()
 }
 
-// Unwrap returns the wrapped error for functions Is and As in standard package errors.
-func (err Logs) Unwrap() error {
-	return err.error
+// Logs returns failed logs from the associated error.
+func (err Logs) Logs() plog.Logs {
+	return err.failed
 }
 
 // Metrics is an error that may carry associated Metrics data for a subset of received data
 // that failed to be processed or sent.
 type Metrics struct {
-	error
-	failed pmetric.Metrics
+	retryable[pmetric.Metrics]
 }
```

Comment:
I think you can do 

```golang
type Metrics retryable[pmetric.Metrics]
```

Reply:
Hmm.. not sure if that does not expose too much about the internal implementation. Let's keep it as is.